include/psa/crypto_struct.h

```diff
@@ -223,6 +223,29 @@ static inline struct psa_key_derivation_s psa_key_derivation_operation_init(
     return v;
 }
 
+struct psa_key_generation_method_s {
+    /* Future versions may add other fields in this structure. */
+    uint32_t flags;
+    uint8_t data[];
+};
+
+/** The default method for key generation or key derivation.
+ *
+ * Calling psa_generate_key_ext() or psa_key_derivation_output_key_ext()
+ * with `method=PSA_KEY_GENERATION_METHOD_INIT` and
+ * `method_length=sizeof(psa_key_generation_method_t)` is equivalent to
```

Comment:
[It's possible that](https://github.com/Mbed-TLS/mbedtls/pull/8815/files#r1486535716) `offsetof(psa_key_generation_method_t, data) < sizeof(psa_key_generation_method_t)` due to padding. As a consequence, `method_length == sizeof(psa_key_generation_method_t)` does not actually mean empty `data` when there is padding. This is problematic. If there is padding and `data` is empty, then the length of the payload in a method object is less than `sizeof(psa_key_generation_method_t)`. But then initializing a structure by copying only `offsetof(psa_key_generation_method_t, data)` bytes is incomplete.

In other words, I'd like to have the ext functions require `method_length >= sizeof(psa_key_generation_method_t)` so that a server can do something like
```
if (method_length < sizeof(psa_key_generation_method_t)) return PSA_ERROR_INVALID_ARGUMENT;
method = malloc(method_length);
memcpy(method, input_buffer, method_length);
```
(assuming the same ABI on the client and server side).

But then on platforms with padding, that makes `data` always be at least `sizeof(psa_key_generation_method_t) - offsetof(psa_key_generation_method_t, data)` bytes long. Whereas if we define empty data to be when `method_length == offsetof(psa_key_generation_method_t, data)`, the server code becomes naively
```
if (method_length < offsetof(psa_key_generation_method_t, data)) return PSA_ERROR_INVALID_ARGUMENT;
method = malloc(method_length);
memcpy(method, input_buffer, method_length);
```
But this is wrong because it doesn't create a valid structure: `*method` would be a buffer overread. So the server code actually needs to fill the end of the structure:
```
if (method_length < offsetof(psa_key_generation_method_t, data)) return PSA_ERROR_INVALID_ARGUMENT;
method = malloc(MAX(method_length, sizeof(psa_key_generation_method_t)));
memcpy(method, input_buffer, method_length);
if (method_length < sizeof(psa_key_generation_method_t)) memset(method + method_length, 0, sizeof(psa_key_generation_method_t) - method_length);
```
Is this ok?

Reply:
Thinking further about it, I think the arguments should be `method` and `method_data_length`. That is, the length parameter should just be the length of the variable part. This removes any confusion between `offsetof(psa_key_generation_method_t, data)` and `sizeof(psa_key_generation_method_t)`.

On client-server implementations, this requires the client and the server to agree on the size of the fixed part of the structure. But that's not a problem. They have to agree anyway to make the API usable: application code needs to fill the structure in the format expected by the server.